File: tic_tac_toe.rb
class Board

  attr_accessor :board, :winner

  def initialize
    @board = [
    ['-','-','-'],
    ['-','-','-'],
    ['-','-','-']
    ]
    @winner
#   board setup:
#   [[0][0]] , [[0],[1]] , [[0][2]]
#   [[1][0]] , [[1],[1]] , [[1][2]]
#   [[2][0]] , [[2],[1]] , [[2][2]]

  end

  def won?
    #[@board[0][0], @board[0][1], @board[0][2]].uniq.count == 1 && (@board[0][0] != '-')

    (0..2).each do |row|
      if @board[row].uniq.count == 1 && @board[row][0] != '-'
        @winner = @board[row][0]
        return true
      end
      if [@board[0][row], @board[1][row], @board[2][row]].uniq.count == 1 && @board[0][row] != '-'
        @winner = @board[0][row]
        return true
      end
    end

    if [@board[0][0], @board[1][1], @board[2][2]].uniq.count == 1 && @board[0][0] != '-' || [@board[0][2], @board[1][1], @board[2][0]].uniq.count == 1 && @board[0][2] != '-'
      @winner = @board[0][0]
      return true
    end


    false
  end


  def empty?(pos)
    # ie. pos upper left corner is [0,0]
    @board[pos.first][pos.last] == '-'
  end

  def place_mark(pos, mark)
      @board[pos.first][pos.last] = mark
  end

  def display_board
    print @board[0]
    puts
    print @board[1]
    puts
    print @board[2]
    puts
    puts
  end

end

class Game

  attr_accessor :player1, :player2, :board

  def initialize(player1, player2)
    @player1 = player1
    @player2 = player2
    @board = Board.new
  end


  def play

    while !@board.won?
      puts "Player 1, what is your move?"
      move = @player1.move

      while !valid?(move)
        puts "Not a valid move, please choose again!"
        move = @player1.move
      end

      @board.place_mark(move, @player1.mark)
      @board.display_board
      if @board.won?
        break
      else
        puts "Player 2, what is your move?"
        move = @player2.move

        while !valid?(move)
          puts "Not a valid move, please choose again!"
          move = @player2.move
        end

        @board.place_mark(move, @player2.mark)
        @board.display_board
      end

    end
  end

  def valid?(move)
    (move.first < 3 && move.first >= 0) && (move.last < 3 && move.last >= 0) && @board.empty?(move)
  end



end

class Player

  attr_accessor :mark, :move

  def initialize

  end

  def move
    [gets.chomp.to_i, gets.chomp.to_i]
  end

end

class HumanPlayer < Player

  def initialize
    @mark = "O"
  end


end

class ComputerPlayer < Player

  def initialize
    @mark = "X"
  end

  def move
    [rand(3), rand(3)]
  end

end

hp = HumanPlayer.new
cp = ComputerPlayer.new

game = Game.new(hp, cp)

game.play
print "#{game.board.winner} is the winner!"

File: tic_tac_toe.js
Board.prototype.isWin = function(mark)  {
  for (var i = 0; i < 3; i++) {
    // rows
    if (this.compareThree([this.grid[i][0],
        this.grid[i][1], this.grid[i][2]]) &&
       (this.grid[i][0] === mark)) {
      return true;
    }

    // cols
    if (this.compareThree([this.grid[0][i],
        this.grid[1][i], this.grid[2][i]]) &&
       (this.grid[0][i] === mark)) {
      return true;
    }
  }
  // diags
  if ((this.compareThree([this.grid[0][0],
      this.grid[1][1], this.grid[2][2]]) ||
     (this.compareThree([this.grid[0][2],
      this.grid[1][1], this.grid[2][0]]))) &&
     (this.grid[1][1] === mark)) {
    console.log(this.grid[1][1] === mark);
    return true;
  }
  return false;
}

Board.prototype.compareThree = function(positions) {
  return (positions[0] === positions[1] && positions[1] === 
          positions[2] && positions[0] != null)
}

Board.prototype.placeMark = function(pos, mark) {
  var x = pos[0];
  var y = pos[1];
  this.grid[x][y] = mark;
}

Board.prototype.validMove = function(pos) {
  var x = pos[0];
  var y = pos[1];
  return ( (this.grid[x] !== undefined) &&
           (this.grid[x][y] !== undefined) &&
           (this.grid[x][y] === null) );
}

Board.prototype.display = function() {
  console.log(this.grid[0]);
  console.log(this.grid[1]);
  console.log(this.grid[2]);
}

Player.prototype.takeTurn = function(board, callback) {
  var that = this;

  board.display();
  console.log();
  console.log("Player: " + that.mark);
  READER.question("What is your move?", function(response) {
    var x = parseInt(response[0]);
    var y = parseInt(response[response.length - 1]);

    if (board.validMove([x, y])) {
      board.placeMark([x, y], that.mark);
      board.switchPlayers(that);
    } else {
      console.log("Invalid move. Please move again!");
      that.takeTurn(board, callback);
    }
    callback();
  });
}

Board.prototype.switchPlayers = function(player) {
  if (player.mark === "X") {
    player.mark = "O";
  } else {
    player.mark = "X";
  }
}

Game.prototype.run = function() {
  var that = this;

  // that.board.display();
  that.currentPlayer.takeTurn(that.board, function () {
    if (that.board.isWin(that.currentPlayer.oppositeMark())) {
      that.board.display();
      console.log("Winner!");
      READER.close();
    } else {
      that.run();
    }
  });
}

Player.prototype.oppositeMark = function() {
  //we need to check the opposite mark in
  //the run method, because we automatically switch players
  //after a valid move
  if (this.mark === "X") {
    return "O";
  } else {
    return "X";
  }
}

function Board () {
  this.grid = [[],[],[]];
  for (var i = 0; i < 3; i++){
    for (var j = 0; j < 3; j++){
      this.grid[i].push(null);
    }
  }
}

function Player(mark) {
  this.mark = mark;
}

function Game() {
  this.board = new Board();
  this.currentPlayer = new Player("X");

  var readline = require('readline');
  READER = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
}

g = new Game();
g.run();